p2panda-rs/src/schema/schema.rs

```diff
@@ -0,0 +1,426 @@
+            Err(err) => Err(SchemaError::MessageError(err)),
+        }
+    }
+
+    /// Create a new UPDATE message validated against this schema
+    #[cfg(not(target_arch = "wasm32"))]
+    pub fn update(&self, id: &str, key_values: Vec<(&str, &str)>) -> Result<Message, SchemaError> {
+        let mut fields = MessageFields::new();
+        let id = Hash::new(id).unwrap();
+
+        for (key, value) in key_values {
+            match fields.add(key, MessageValue::Text(value.into())) {
+                Ok(_) => Ok(()),
+                Err(err) => Err(SchemaError::InvalidSchema(err.to_string())),
+            }?;
+        }
+
+        match self.validate_message(serde_cbor::to_vec(&fields.clone()).unwrap()) {
+            Ok(_) => Ok(()),
+            Err(err) => Err(SchemaError::ValidationError(err.to_string())),
+        }?;
+
+        match Message::new_update(self.schema_hash(), id, fields) {
+            Ok(hash) => Ok(hash),
+            Err(err) => Err(SchemaError::InvalidSchema(err.to_string())),
+        }
+    }
+
+    /// Validate a message against this user schema
+    #[cfg(not(target_arch = "wasm32"))]
+    pub fn validate_message(&self, bytes: Vec<u8>) -> Result<(), SchemaError> {
+        match validate_cbor_from_slice(&format!("{}", self), &bytes) {
+            Err(cbor::Error::Validation(err)) => {
+                let err = err
+                    .iter()
+                    .map(|fe| format!("{}: \"{}\"", fe.cbor_location, fe.reason))
+                    .collect::<Vec<String>>()
+                    .join(", ");
+
+                Err(SchemaError::InvalidSchema(err))
+            }
+            Err(cbor::Error::CBORParsing(_err)) => Err(SchemaError::InvalidCBOR),
+            Err(cbor::Error::CDDLParsing(err)) => {
+                panic!("Parsing CDDL error: {}", err);
+            }
+            _ => Ok(()),
+        }
+    }
+}
+
+impl fmt::Display for Schema {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.schema_string)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::hash::Hash;
+    use crate::message::{Message, MessageFields, MessageValue};
+    use crate::schema::{
+        Schema, SchemaBuilder, Type, PERSON_SCHEMA, PERSON_SCHEMA_HASH, USER_SCHEMA,
+        USER_SCHEMA_HASH,
+    };
+
+    #[test]
+    pub fn schema_builder() {
+        // Instanciate new empty schema named "person"
+        let mut person = SchemaBuilder::new("person".to_owned());
+
+        // Add two message fields to the schema
+        person
+            .add_message_field("name".to_owned(), Type::Tstr)
+            .unwrap();
+        person
+            .add_message_field("age".to_owned(), Type::Int)
+            .unwrap();
+
+        // Create a new "person" message
+        let mut me = MessageFields::new();
+        me.add("name", MessageValue::Text("Sam".to_owned()))
+            .unwrap();
+        me.add("age", MessageValue::Integer(35)).unwrap();
```

Comment:
Beautiful! I really like the `SchemaBuilder` pattern here. Will be nice to introduce this to `Message` as well (`MessageBuilder`) :+1:

Maybe we can introduce an `Instance` struct one day which you can also use to validate Schemas against.

Reply:
Ohh, thanks! I was struggling to make sense of the all in one schema struct, once it's split into 2 it becomes really simple and clear 👍👌

Nice, I'm looking forward to our Instance struct, it could appear at any moment 😁